File: stacks/fastapi/app/routers/sales.py
"""売上集計API。`GET /api/sales/summary`。

フィールド形について（`coordination/qa/lane-d.md` D-5）:
`spec/openapi.yaml` の `SalesSummary` は `from`/`to` を必須クエリにし、
`group_by` で1軸だけ返す形だが、共通テスト（`tests/checks.py` の検算1）は
クエリ無しで叩き、`by_category`/`by_staff`/`by_date` の3軸を**同時に**、
かつ分類別の各行に `share_pct` を求めている。共通テストが判定の実体なので、
そちらの形を主に返す（`from`/`to` は任意にし、渡されれば絞り込みに使う）。

3方向一致・構成比の計算規則は `spec/acceptance.md`「検算1」「数値の規則」のとおり:
- 対象は `Billing.status = confirmed` のみ
- `unit_price` 未設定の明細はどの合計にも入れない
- 担当は `staff_id`（`cashier_staff_id` ではない）
- 構成比は最大剰余法（`app/billing_calc.allocate_share_pct`）
"""

from __future__ import annotations

from collections import defaultdict
from decimal import Decimal

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app import fixtures, models
from app.billing_calc import allocate_share_pct
from app.db import get_db

router = APIRouter(prefix="/api", tags=["api-billing"])


def _category_major(price_code: str) -> str | None:
    item = fixtures.price_item_by_code(price_code)
    if not item:
        return None
    return item.get("category_major") or item.get("category")


def compute_summary(db: Session, date_from: str | None = None, date_to: str | None = None) -> dict:
    """`/sales` 画面（`app/routers/front.py`）とAPIが共有する集計本体。

    2箇所に書き写すと片方だけ直し忘れる事故が起きるので、ここへ1本化した
    （`_visit_macro.html` を画面と印刷で共有したのと同じ考え方）。
    """
    billings = db.query(models.Billing).filter(models.Billing.status == "confirmed").all()

    by_cat: dict[str, Decimal] = defaultdict(Decimal)
    by_staff: dict[int, Decimal] = defaultdict(Decimal)
    by_date: dict[str, Decimal] = defaultdict(Decimal)
    excluded_total = 0
    total = Decimal(0)

    for b in billings:
        day = b.billed_on.isoformat()
        if date_from and day < date_from:
            continue
        if date_to and day > date_to:
            continue
        for d in b.details:
            if d.unit_price is None:
                excluded_total += 1
                continue
            amount = Decimal(str(d.quantity)) * Decimal(d.unit_price)
            total += amount
            by_date[day] += amount
            by_staff[b.staff_id] += amount
            cat = _category_major(d.price_code)
            if cat is not None:
                by_cat[cat] += amount

    cat_rows = allocate_share_pct(list(by_cat.items()))

    return {
        # 共通テストが読む名前。
        "total_net_amount": int(total),
        "by_category": [
            {"category": r.key, "net_amount": r.net_amount, "share_pct": r.share_pct}
            for r in cat_rows
        ],
        "by_staff": [
            {"staff_id": k, "net_amount": int(v)} for k, v in by_staff.items()
        ],
        "by_date": [
            {"date": k, "net_amount": int(v)} for k, v in by_date.items()
        ],
        # openapi.yaml 側の名前（互換のため併記）。
        "total": int(total),
        "total_amount": int(total),
        "excluded_detail_count_total": excluded_total,
        "from": date_from,
        "to": date_to,
    }


@router.get("/sales/summary")
def sales_summary(
    db: Session = Depends(get_db),
    date_from: str | None = Query(None, alias="from"),
    date_to: str | None = Query(None, alias="to"),
):
    return compute_summary(db, date_from, date_to)

File: stacks/nextjs/src/lib/sales.ts
/**
 * Sales summary (売上集計, 新規画面) — 領域3｜会計・売上.
 *
 * `spec/screens.md`「17. 売上集計」and `spec/acceptance.md`「検算1：売上の3方向
 * 一致」. The three breakdowns (分類別・担当別・日別) are three different
 * groupings of *the same* priced, confirmed detail rows, run through the
 * same `computeBillingTotals` used everywhere else in this area -- so their
 * subtotals sum to the same grand total by construction, not by careful
 * arithmetic that could drift.
 *
 * `rulings.md` #1/#3: 担当軸は `Billing.staff_id`（`cashier_staff_id` ではない）。
 * 消費税は伝票単位・按分しない。検算1の対象は税抜（`subtotal`）のみ -- each
 * row's own `tax_amount`/`total` (required by `spec/openapi.yaml`
 * `SalesSummary`) are filled in by applying the same floor-once rule to that
 * row's own bucket of lines, which keeps every row internally consistent
 * (`subtotal + tax_amount = total`) without needing to apportion a real
 * billing's tax across groups. Nothing here claims that per-row tax is
 * exact for the 分類 axis in the way a real invoice's tax would be --
 * 検算1 deliberately does not check it (see acceptance.md's note under
 * 検算1 about why tax is excluded from the 3-way comparison).
 */
import { getDb, rows, row } from './db';
import { computeBillingTotals, largestRemainderPercent, type DetailForTotals } from './money';
import { loadPriceItems } from './price-items.ts';

let categoryCache: Map<string, string> | undefined;
function categoryByPriceCode(): Map<string, string> {
  if (!categoryCache) {
    categoryCache = new Map(loadPriceItems().map((p) => [p.price_code, p.category_major]));
  }
  return categoryCache;
}

type Priced = DetailForTotals & { price_code: string; staff_id: number | null; billed_on: string };

export type SummaryRow = {
  key: string;
  billing_count: number;
  net_amount: number; // subtotal, 税抜 (taxable_subtotal + nontaxable_subtotal)
  tax_amount: number;
  total: number;
  excluded_detail_count: number;
  share_pct?: number | null;
};

export type SalesSummaryResult = {
  from: string;
  to: string;
  net_amount_total: number;
  tax_amount_total: number;
  total: number;
  excluded_detail_count_total: number;
  by_category: SummaryRow[];
  by_staff: SummaryRow[];
  by_date: SummaryRow[];
};

function taxRate(): number {
  return row<{ tax_rate: number }>(getDb().prepare('SELECT tax_rate FROM clinic LIMIT 1'))!.tax_rate;
}

function bucket<K extends string>(details: Priced[], keyOf: (d: Priced) => K, billingCountOf: (details: Priced[]) => number, rate: number): Map<K, SummaryRow> {
  const groups = new Map<K, Priced[]>();
  for (const d of details) {
    const k = keyOf(d);
    (groups.get(k) ?? groups.set(k, []).get(k)!).push(d);
  }
  const out = new Map<K, SummaryRow>();
  for (const [k, ds] of groups) {
    const totals = computeBillingTotals(ds, rate);
    out.set(k, {
      key: k,
      billing_count: billingCountOf(ds),
      net_amount: totals.taxable_subtotal + totals.nontaxable_subtotal,
      tax_amount: totals.tax_amount,
      total: totals.total,
      excluded_detail_count: totals.excluded_detail_count,
    });
  }
  return out;
}

/**
 * `Billing.status = confirmed` only (`draft` never counts toward sales,
 * `spec/screens.md` 17). `unit_price === null` rows are excluded by
 * `computeBillingTotals` itself, and every row here goes through it, so
 * exclusion is uniform across all three axes and the total.
 */
export function computeSalesSummary(from: string, to: string): SalesSummaryResult {
  const db = getDb();
  const rate = taxRate();

  const billings = rows<{ id: number; staff_id: number | null; billed_on: string }>(
    db.prepare("SELECT id, staff_id, billed_on FROM billing WHERE status = 'confirmed' AND billed_on >= ? AND billed_on <= ?"),
    from,
    to,
  );
  const billingById = new Map(billings.map((b) => [b.id, b]));

  const allDetails = billings.length === 0
    ? []
    : rows<{ billing_id: number; price_code: string; quantity: number; unit_price: number | null; is_taxable: number }>(
        db.prepare(
          `SELECT billing_id, price_code, quantity, unit_price, is_taxable FROM billing_detail WHERE billing_id IN (${billings.map(() => '?').join(',')})`,
        ),
        ...billings.map((b) => b.id),
      );

  const priced: Priced[] = allDetails.map((d) => {
    const b = billingById.get(d.billing_id)!;
    return {
      quantity: d.quantity,
      unit_price: d.unit_price,
      is_taxable: !!d.is_taxable,
      price_code: d.price_code,
      staff_id: b.staff_id,
      billed_on: b.billed_on,
    };
  });

  const excludedTotal = allDetails.filter((d) => d.unit_price === null).length;

  const categoryMap = categoryByPriceCode();
  const byCategory = bucket(
    priced,
    (d) => categoryMap.get(d.price_code) ?? 'unknown',
    (ds) => new Set(ds.map((d) => `${d.staff_id}:${d.billed_on}`)).size, // not a meaningful billing count across mixed billings; see rows below for the real count
    rate,
  );
  const byStaff = bucket(priced, (d) => String(d.staff_id), () => 0, rate);
  const byDate = bucket(priced, (d) => d.billed_on, () => 0, rate);

  // billing_count per group: count distinct billings whose OWN axis value
  // matches the group (staff_id / billed_on are billing-level, so this is
  // exact; for category, a billing counts toward every category it has a
  // priced line in).
  const billingCountByCategory = new Map<string, Set<number>>();
  for (const [id, b] of billingById) {
    const cats = new Set(
      allDetails.filter((d) => d.billing_id === id && d.unit_price !== null).map((d) => categoryMap.get(d.price_code) ?? 'unknown'),
    );
    for (const c of cats) (billingCountByCategory.get(c) ?? billingCountByCategory.set(c, new Set()).get(c)!).add(id);
  }
  for (const [k, r] of byCategory) r.billing_count = billingCountByCategory.get(k)?.size ?? 0;

  const billingCountByStaff = new Map<string, Set<number>>();
  const billingCountByDate = new Map<string, Set<number>>();
  for (const [id, b] of billingById) {
    const hasPriced = allDetails.some((d) => d.billing_id === id && d.unit_price !== null);
    if (!hasPriced) continue;
    const sKey = String(b.staff_id);
    const dKey = b.billed_on;
    (billingCountByStaff.get(sKey) ?? billingCountByStaff.set(sKey, new Set()).get(sKey)!).add(id);
    (billingCountByDate.get(dKey) ?? billingCountByDate.set(dKey, new Set()).get(dKey)!).add(id);
  }
  for (const [k, r] of byStaff) r.billing_count = billingCountByStaff.get(k)?.size ?? 0;
  for (const [k, r] of byDate) r.billing_count = billingCountByDate.get(k)?.size ?? 0;

  const grand = computeBillingTotals(priced, rate);
  const netTotal = grand.taxable_subtotal + grand.nontaxable_subtotal;

  const shares = largestRemainderPercent(
    [...byCategory.entries()].map(([key, r]) => ({ key, value: r.net_amount })),
    netTotal,
  );
  const categoryRows = [...byCategory.values()]
    .map((r) => ({ ...r, share_pct: shares?.get(r.key) ?? null }))
    .sort((a, b) => a.key.localeCompare(b.key, 'ja'));

  return {
    from,
    to,
    net_amount_total: netTotal,
    tax_amount_total: grand.tax_amount,
    total: grand.total,
    excluded_detail_count_total: excludedTotal,
    by_category: categoryRows,
    by_staff: [...byStaff.values()].sort((a, b) => a.key.localeCompare(b.key)),
    by_date: [...byDate.values()].sort((a, b) => a.key.localeCompare(b.key)),
  };
}
